doc/content/toolstack/features/NUMA/index.md

```diff
@@ -62,39 +62,59 @@ The Xen scheduler supports 2 kinds of constraints:
 * hard pinning: a vCPU may only run on the specified set of pCPUs and nowhere else
 * soft pinning: a vCPU is *preferably* run on the specified set of pCPUs, but if they are all busy then it may run elsewhere
 
-The former is useful if you want strict separation, but it can potentially leave part of the system idle while another part is bottlenecked with lots of vCPUs all competing for the same limited set of pCPUs.
+Hard pinning can be used to partition the system. But, it can potentially leave part of the system idle while another part is bottlenecked by many vCPUs competing for the same limited set of pCPUs.
 
-Xen does not migrate workloads between NUMA nodes on its own (the Linux kernel does), although it is possible to achieve a similar effect with explicit migration.
-However migration introduces additional delays and is best avoided for entire VMs.
+Xen does not migrate workloads between NUMA nodes on its own (the Linux kernel can). Although, it is possible to achieve a similar effect with explicit migration.
+However, migration introduces additional delays and is best avoided for entire VMs.
 
-The latter (soft pinning) is preferred: running a workload now, even on a potentially suboptimal pCPU (higher NUMA latency) is still better than not running it at all and waiting until a pCPU is freed up.
+Therefore, soft pinning is preferred: Running on a potentially suboptimal pCPU that uses remote memory could still be better than not running it at all until a pCPU is free to run it.
```

Comment:
This was the initial idea, but I think it has been proven partially wrong due to how Xen's credit1 scheduler works that can only take local decisions: it won't realise that it is running all the vcpus in the suboptimal place and that migrating some VMs to other CPUs would restore them according to the soft affinity mask.
That is not a simple bug to fix, might require a more fundamental design change of the scheduler.

Reply:
Yes, that combination of hard and soft pinning would be great!

An example where this could be applied would be to use hard-pinning for the CPU packages(sockets) (in the example CPU in `new-docs`): NUMA distance >= 21) and soft-pinning could them be used below it like you said.

In the example CPU in `new-docs`, the 2nd level NUMA would be the NUMA distance between the two NUMA nodes within each CPU package(socket) where the NUMA distance is only 11 due to the split L3 cache (and below that the memory controllers/busses) of each socket.

As you mentioned in the approval comment, this update can be planned for a future PR.